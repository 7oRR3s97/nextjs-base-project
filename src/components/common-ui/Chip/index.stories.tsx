import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import ChipComponent from './index';

export default {
  title: 'CommonUI/Chip',
  component: ChipComponent,
} as ComponentMeta<typeof ChipComponent>;

const Template: ComponentStory<typeof ChipComponent> = (args) => (
  <ChipComponent {...args} />
);

export const Primary = Template.bind({});
Primary.args = {
  label: 'chip',
  error: false,
  success: false,
  warning: false,
};

export const Error = Template.bind({});
Error.args = {
  label: 'chip',
  error: true,
};

export const Warning = Template.bind({});
Warning.args = {
  label: 'chip',
  warning: true,
};

export const Success = Template.bind({});
Success.args = {
  label: 'chip',
  success: true,
};
