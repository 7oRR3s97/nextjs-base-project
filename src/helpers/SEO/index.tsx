import { useTranslation } from 'next-i18next';

export interface Metadata {
  title: string;
  image: string;
  description: string;
  ogType: string;
  url: string;
  page: string;
  locale: string;
  params?: { [key: string]: string | string[] };
}

// TODO: Alterar dados para SEO
const useSeo = (props: Partial<Metadata>) => {
  const { t: userPageT } = useTranslation('seo-user-page');
  const { t: homePageT } = useTranslation('seo-home-page');
  if (props.page === 'user')
    return {
      title: userPageT('title'),
      image:
        'https://assets.website-files.com/613a4aeb1cd38a810f6a1448/61689732f4b0536181967320_Sem-Ti%CC%81tulo-1.png',
      description: userPageT('description'),
      url: `https://isaac.com.br/minhaconta`,
      ogType: 'website',
    };
  return {
    title: homePageT('title'),
    image:
      'https://assets.website-files.com/613a4aeb1cd38a810f6a1448/61689732f4b0536181967320_Sem-Ti%CC%81tulo-1.png',
    description: homePageT('description'),
    url: 'https://isaac.com.br',
    ogType: 'website',
  };
};

export default useSeo;
