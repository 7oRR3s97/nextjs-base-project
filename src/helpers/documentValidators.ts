import range from 'lodash/range';

export const CPFValidator = (data: string) => {
  const cpf = data.replace(/\.|-/g, '');

  if (
    cpf.length !== 11 ||
    cpf === '00000000000' ||
    cpf === '11111111111' ||
    cpf === '22222222222' ||
    cpf === '33333333333' ||
    cpf === '44444444444' ||
    cpf === '55555555555' ||
    cpf === '66666666666' ||
    cpf === '77777777777' ||
    cpf === '88888888888' ||
    cpf === '99999999999'
  )
    return false;

  // Validate first digit
  const firstDigitSum = range(0, 9).reduce((sum, item) => {
    sum += parseInt(cpf.charAt(item), 10) * (10 - item);
    return sum;
  }, 0);

  const firstDigitRemainder = 11 - (firstDigitSum % 11);
  const firstDigit = [10, 11].includes(firstDigitRemainder)
    ? 0
    : firstDigitRemainder;
  if (firstDigit !== parseInt(cpf.charAt(9), 10)) return false;

  // Validate second digit
  const secondDigitSum = range(0, 10).reduce((sum, item) => {
    sum += parseInt(cpf.charAt(item), 10) * (11 - item);
    return sum;
  }, 0);

  const secondDigitRemainder = 11 - (secondDigitSum % 11);
  const secondDigit = [10, 11].includes(secondDigitRemainder)
    ? 0
    : secondDigitRemainder;
  if (secondDigit !== parseInt(cpf.charAt(10), 10)) return false;

  return true;
};

export const CNPJValidator = (data: string) => {
  const cnpj = data.replace(/\.|\/|-/g, '');
  const size = cnpj.length - 2;
  const digits = cnpj.substring(size);
  const firstNumbers = cnpj.substring(0, size);

  if (
    cnpj.length !== 14 ||
    cnpj === '00000000000000' ||
    cnpj === '11111111111111' ||
    cnpj === '22222222222222' ||
    cnpj === '33333333333333' ||
    cnpj === '44444444444444' ||
    cnpj === '55555555555555' ||
    cnpj === '66666666666666' ||
    cnpj === '77777777777777' ||
    cnpj === '88888888888888' ||
    cnpj === '99999999999999'
  )
    return false;

  // Valida DVs
  const { sum: firstSum } = range(size, 0, -1).reduce(
    ({ sum, position }, item) => {
      sum += parseInt(firstNumbers.charAt(size - item), 10) * position;
      position -= 1;
      if (position < 2) position = 9;
      return { sum, position };
    },
    { sum: 0, position: size - 7 }
  );

  const firstResult = firstSum % 11 < 2 ? 0 : 11 - (firstSum % 11);
  if (firstResult !== parseInt(digits.charAt(0), 10)) return false;

  const currentSize = size + 1;
  const currentNumbers = cnpj.substring(0, currentSize);

  const { sum: secondSum } = range(currentSize, 0, -1).reduce(
    ({ sum, position }, item) => {
      sum += parseInt(currentNumbers.charAt(currentSize - item), 10) * position;
      position -= 1;
      if (position < 2) position = 9;
      return { sum, position };
    },
    { sum: 0, position: currentSize - 7 }
  );

  const secondResult = secondSum % 11 < 2 ? 0 : 11 - (secondSum % 11);
  if (secondResult !== parseInt(digits.charAt(1), 10)) return false;

  return true;
};

export const NIFValidator = (documentData: string) => {
  if (
    documentData.length < 9 ||
    (documentData.substring(0, 1) !== '1' && // pessoa singular
      documentData.substring(0, 1) !== '2' && // pessoa singular
      documentData.substring(0, 1) !== '3' && // pessoa singular
      documentData.substring(0, 2) !== '45' && // pessoa singular não residente
      documentData.substring(0, 1) !== '5' && // pessoa colectiva
      documentData.substring(0, 1) !== '6' && // administração pública
      documentData.substring(0, 2) !== '70' && // herança indivisa
      documentData.substring(0, 2) !== '71' && // pessoa colectiva não residente
      documentData.substring(0, 2) !== '72' && // fundos de investimento
      documentData.substring(0, 2) !== '77' && // atribuição oficiosa
      documentData.substring(0, 2) !== '79' && // regime excepcional
      documentData.substring(0, 1) !== '8' && // empresário em nome individual (extinto)
      documentData.substring(0, 2) !== '90' && // condominios e sociedades irregulares
      documentData.substring(0, 2) !== '91' && // condominios e sociedades irregulares
      documentData.substring(0, 2) !== '98' && // não residentes
      documentData.substring(0, 2) !== '99') // sociedades civis
  )
    return false;

  const nifWithoutLastDigit = documentData.slice(0, documentData.length - 1);

  const remainder =
    range(9, 1, -1).reduce((sum, item) => {
      sum += parseInt(documentData.charAt(9 - item), 10) * item;
      return sum;
    }, 0) % 11;

  if ([0, 1].includes(remainder))
    return (documentData === `${nifWithoutLastDigit}0`)!!;
  return (documentData === `${nifWithoutLastDigit}${11 - remainder}`)!!;
};

export const documentValidator = (
  documentData: string,
  documentType: string,
  country: string
) => {
  if (country.toLowerCase() === 'portugal') return NIFValidator(documentData);
  if (country.toLowerCase() === 'brasil') {
    if (documentType.toLowerCase() === 'cpf') return CPFValidator(documentData);
    if (documentType.toLowerCase() === 'cnpj')
      return CNPJValidator(documentData);
    return false;
  }
  return false;
};
