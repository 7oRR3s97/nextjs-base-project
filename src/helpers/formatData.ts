import accounting from 'accounting-js';

const moneyFormat = (currency?: string) => {
  if (currency === 'BRL') return ['R$', '%s%v', '.', ','];
  if (currency === 'USD') return ['US$', '%s%v', ',', '.'];
  if (currency === 'EUR') return ['â‚¬', '%v%s', ',', '.'];
  return ['R$', '%s%v', '.', ','];
};

export const formatMoney = (value: number, currency?: string) => {
  const [currencySymbol, currencyFormat, currencyThousand, currencyDecimal] =
    moneyFormat(currency);

  const money = accounting.formatMoney(value, {
    symbol: currencySymbol,
    format: currencyFormat,
    precision: 2,
    thousand: currencyThousand,
    decimal: currencyDecimal,
  });

  return money;
};

export const formatKBRT = (value: number) => {
  const KBRT = accounting.formatNumber(value, {
    thousand: '.',
    decimal: ',',
  });

  return `${KBRT} KBRT`;
};

export const formatDate = (date: string) => {
  const dateInstance = new Date(date);
  const day = dateInstance.getDate();
  const month = dateInstance.getMonth() + 1;
  const formattedDate = `${day < 10 ? `0${day}` : day}/${
    month < 10 ? `0${month}` : month
  }/${dateInstance.getFullYear()}`;

  return formattedDate;
};

export const formatHour = (date: string) => {
  const dateInstance = new Date(date);

  const hours = dateInstance.getHours();
  const minutes = dateInstance.getMinutes();
  const seconds = dateInstance.getSeconds();
  const formattedHour = `${hours < 10 ? `0${hours}` : hours}:${
    minutes < 10 ? `0${minutes}` : minutes
  }:${seconds < 10 ? `0${seconds}` : seconds}`;

  return formattedHour;
};

export const formatDateHour = (date: string) => {
  const formattedDate = formatDate(date);
  const formattedHour = formatHour(date);

  return `${formattedHour} - ${formattedDate}`;
};
