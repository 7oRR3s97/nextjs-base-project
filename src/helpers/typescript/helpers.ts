export const objectKeys = <TObj>(obj: TObj): (keyof TObj)[] => {
  return Object.keys(obj) as (keyof TObj)[];
};

export const getDynamicValueFromObject = <TObj>(
  obj: TObj,
  index: string
): TObj[keyof TObj] | undefined => {
  if (!Object.keys(obj).includes(index)) return;
  return obj[index as keyof TObj];
};
