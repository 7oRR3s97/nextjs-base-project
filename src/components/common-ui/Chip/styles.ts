import { styled } from '@mui/material/styles';
import Chip from '@mui/material/Chip';

export const ChipComponent = styled(Chip)(({ theme }) => ({
  '&.MuiChip-colorSuccess': {
    color: theme.palette.success.dark,
    backgroundColor: theme.palette.success.light,
  },
  '&.MuiChip-colorWarning': {
    color: theme.palette.warning.dark,
    backgroundColor: theme.palette.warning.light,
  },
  '&.MuiChip-colorError': {
    color: theme.palette.error.dark,
    backgroundColor: theme.palette.error.light,
  },
}));
