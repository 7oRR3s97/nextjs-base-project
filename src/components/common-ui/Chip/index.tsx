import React from 'react';
import type { ChipProps } from '@mui/material/Chip';

import { ChipComponent } from './styles';

interface Props extends ChipProps {
  label?: string;
  success?: boolean;
  error?: boolean;
  warning?: boolean;
}

const Chip: React.FC<Props> = ({
  label,
  success,
  error,
  warning,
  ...props
}) => {
  const color = React.useMemo(() => {
    if (error) return 'error';
    if (warning) return 'warning';
    if (success) return 'success';
    return 'primary';
  }, [error, warning, success]);

  return <ChipComponent {...props} label={label} color={color} />;
};

export default Chip;
