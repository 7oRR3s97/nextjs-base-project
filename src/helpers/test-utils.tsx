import React from 'react';
import {
  render as defaultRender,
  renderHook as defaultRendesHook,
  RenderHookOptions,
  RenderHookResult,
  act,
  RenderOptions,
} from '@testing-library/react';
import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import { I18nextProvider } from 'react-i18next';
import { ThemeProvider } from '@mui/material/styles';
import { Provider } from 'react-redux';

import { store } from 'redux/store';

import theme from 'styles/theme';
import QueryClientProvider from 'components/Providers/queryClientProvider';

const ns = [
  'account-data',
  'auth-alerts',
  'common',
  'edit-data',
  'home-page-callout',
  'login',
  'menu',
  'seo-home-page',
  'seo-user-page',
];
const supportedLngs = ['pt-br', 'en'];

i18n.use(initReactI18next).init({
  fallbackLng: 'en',
  debug: false,
  interpolation: {
    escapeValue: false,
  },
  defaultNS: 'common',
  ns,
  supportedLngs,
  react: {
    useSuspense: false,
  },
  resources: {
    en: {},
    'pt-br': {},
    pt: {},
  },
});

const Wrapper: React.FC = ({ children }) => {
  return (
    <Provider store={store}>
      <QueryClientProvider>
        <I18nextProvider i18n={i18n}>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </I18nextProvider>
      </QueryClientProvider>
    </Provider>
  );
};

const render = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper' | 'store'>
) => {
  return defaultRender(ui, { wrapper: Wrapper, ...options });
};

const renderHook = <Result, Props>(
  render: (initialProps: Props) => Result,
  options?: Omit<RenderHookOptions<Props>, 'wrapper'>
): RenderHookResult<Result, Props> => {
  return defaultRendesHook(render, { wrapper: Wrapper, ...options });
};

const advanceTimersByTime = async (time: number) =>
  await act(async () => {
    jest.advanceTimersByTime(time);
  });

/* istanbul ignore next */
export * from '@testing-library/react';
export { render, renderHook, advanceTimersByTime };
