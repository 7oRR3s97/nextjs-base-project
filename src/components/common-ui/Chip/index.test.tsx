import React from 'react';
import { render, screen, cleanup } from 'helpers/test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import { composeStories } from '@storybook/testing-react';
import Chip from './index';

import * as ChipStories from './index.stories';

expect.extend(toHaveNoViolations);

describe('Chip', () => {
  afterEach(() => {
    cleanup();
  });

  const stories = Object.values(composeStories(ChipStories)).map((Story) => {
    return {
      name: Story.storyName!,
      Story,
    };
  });

  it.each(stories)(
    'should have no accessibility error in $name chip story',
    async ({ Story }) => {
      const { container } = render(<Story />);
      const AxeResults = await axe(container);
      expect(AxeResults).toHaveNoViolations();
    }
  );

  it.each(stories)('should render a $name chip from story', ({ Story }) => {
    render(<Story data-testid="chip-component" />);

    const chip = screen.queryByTestId('chip-component');
    expect(chip).toMatchSnapshot();
    expect(chip).toBeInTheDocument();
  });

  it('should render a primary chip', () => {
    render(<Chip data-testid="chip-component" label="chip" />);

    const chip = screen.queryByTestId('chip-component');
    expect(chip).toMatchSnapshot();
    expect(chip).toBeInTheDocument();
  });

  it('should render an error chip', () => {
    render(<Chip data-testid="chip-component" label="chip" error />);

    const chip = screen.queryByTestId('chip-component');
    expect(chip).toMatchSnapshot();
    expect(chip).toBeInTheDocument();
  });

  it('should render a warning chip', () => {
    render(<Chip data-testid="chip-component" label="chip" warning />);

    const chip = screen.queryByTestId('chip-component');
    expect(chip).toMatchSnapshot();
    expect(chip).toBeInTheDocument();
  });

  it('should render a success chip', () => {
    render(<Chip data-testid="chip-component" label="chip" success />);

    const chip = screen.queryByTestId('chip-component');
    expect(chip).toMatchSnapshot();
    expect(chip).toBeInTheDocument();
  });
});
