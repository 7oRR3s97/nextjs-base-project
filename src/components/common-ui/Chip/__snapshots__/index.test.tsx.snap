// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip should render a Error chip from story 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorError MuiChip-filledError css-q7go4k-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;

exports[`Chip should render a Primary chip from story 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-filledPrimary css-9ei9a0-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;

exports[`Chip should render a Success chip from story 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorSuccess MuiChip-filledSuccess css-1pptgcf-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;

exports[`Chip should render a Warning chip from story 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorWarning MuiChip-filledWarning css-3hdox0-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;

exports[`Chip should render a primary chip 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-filledPrimary css-9ei9a0-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;

exports[`Chip should render a success chip 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorSuccess MuiChip-filledSuccess css-1pptgcf-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;

exports[`Chip should render a warning chip 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorWarning MuiChip-filledWarning css-3hdox0-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;

exports[`Chip should render an error chip 1`] = `
<div
  class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorError MuiChip-filledError css-q7go4k-MuiChip-root"
  data-testid="chip-component"
>
  <span
    class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
  >
    chip
  </span>
</div>
`;
